src/pso.c

```diff
@@ -0,0 +1,204 @@
+                          size_t swarm_size, size_t dim,
+                          const double* const min_positions,
+                          const double* const max_positions) {
+  double* u = (double *)malloc(swarm_size*dim*sizeof(double));
+  u = rand_init(swarm_size,dim,min_positions,max_positions);
+  double* v = (double *)malloc(swarm_size*dim*sizeof(double));
+  for (size_t particle=0; particle < swarm_size; particle++){
+    for (size_t d=0; d<dim; d++){
+      size_t idx = (particle*dim) + d;
+      v[idx] = 0.25*(u[idx] - x[idx]);
+    }
+  }
+  return v;
+}
+
+  // sorts the population in decreasing order of fitness /
+  // i.e pop[0] posize_ts to best location yet.
+size_t best_fitness(double* fitness,size_t dim, size_t swarm_size) {
+  // find minimum of fitness
+  // set index of mimima as minima_idx;
+  // global_best = x+(dim*minima_idx);
+  size_t minima_idx = 0;
+  double best_yet = fitness[0]; // global best at timestep t+1
+  for (size_t p=1;p<swarm_size;p++){
+    if (fitness[p]<best_yet) {
+      best_yet = fitness[p];
+      minima_idx = p;
+    }
+  }
+  // global_best_position = local_best_position+(dim*minima_idx);
+  return minima_idx;
+}
+
+  // limit max and min velocity values to ensure no boundary overflow
+void generate_vel_limit(const double* min_positions,
+                        const double* max_positions,
+                        double* min_vel,
+                        double* max_vel,
+                        size_t dim) {
+  for (size_t d=0;d<dim;d++){
+    min_vel[d] = min_positions[d]/5;
+    max_vel[d] = max_positions[d]/5;
+  }
+}
+
+  // updates velocities, works on array returned by gen_velocity
+double* update_velocity(double * v, double* x, double * y,
+                        double* best, size_t swarm_size, size_t dim,
+                        const double* const min_vel,
+                        const double* const max_vel) {
+  double* vnew = (double*)malloc(dim*swarm_size*sizeof(double));
+  for (size_t particle=0; particle < swarm_size; particle++){
+    for (size_t d=0; d<dim; d++){
+      size_t idx = (particle*dim) + d;
+      vnew[idx] = INERTIA*v[idx] + COG*random_0_to_1()*(y[idx] - x[idx]) +
+      SOCIAL*random_0_to_1()*(best[d] * x[idx]);
+      vnew[idx] = min(max(min_vel[d],vnew[idx]),max_vel[d]);
+    }
+  }
+  return vnew;
+}
+
+void update_position(double* x, double* v,
+                    size_t swarm_size, size_t dim,
+                    const double* const min_positions,
+                    const double* const max_positions) {
+  for (size_t particle=0; particle < swarm_size; particle++){
+    for (size_t d=0; d<dim; d++){
+      size_t idx = (particle*dim) + d;
+      x[idx] += v[idx];
+      if (x[idx] >= max_positions[d] || x[idx] <= min_positions[d]){
+        v[idx] = -v[idx];
+      }
+      x[idx] = min(max(min_positions[d],x[idx]),max_positions[d]);
+      //TODO: PROJECT ON BOUNDS - RATHER DO A BOUNCE THAN A SIMPLE PROJECT

```

Comment:
would be cool, please setup a github issue rather than leaving a TODO in reviewed code